traffic_ops/traffic_ops_golang/deliveryservice/keys.go

```diff
@@ -268,29 +273,107 @@ WHERE r.pattern = $2
 	return xmlID, true, nil
 }
 
+// decode the private key
+// check for proper algorithm.
+// check for correct number of keys
+// return private key object, cleaned private key PEM, or any errors.
+func decodePrivateKey(pemPrivateKey string) (*rsa.PrivateKey, string, error) {
+
+	// Remove any white space before decoding
+	trimmedPrivateKey := strings.TrimSpace(pemPrivateKey)
+
+	// Check for proper key count before attempting to decode.
+	blockCount := strings.Count(trimmedPrivateKey, "\n-----END")
+	if blockCount < 1 {
+		return nil, "", errors.New("no private keys found")
+	} else if blockCount > 1 {
+		return nil, "", errors.New("multiple private keys found")
+	}
+
+	// Attempt to decode pem encoded text into PEM block.
+	block, _ := pem.Decode([]byte(trimmedPrivateKey))
+	if block == nil {
+		return nil, "", errors.New("could not decode pem-encoded private key")
+	}
+
+	// Check for encrypted keys or other unsupported key types
+	if strings.Contains(block.Type, "ENCRYPTED") {
+		return nil, "", errors.New("encrypted private key not supported")
+	}
+
+	// Check block headers for encryption.
+	for _, value := range block.Headers {
+		if strings.Contains(value, "ENCRYPTED") {
+			return nil, "", errors.New("encrypted private key not supported")
+		}
+	}
+
+	// Attempt to parse PEM block as a PKCS#1 or PKCS#8 formatted RSA Private Key.
+	if privateKey, err := x509.ParsePKCS8PrivateKey(block.Bytes); privateKey != nil && err == nil {
```

Comment:
I think instead of `if thing, err := doThing(); thing != nil && err == nil {...` you might want to do:

```go
thing, err := doThing()
if err != nil {
    return nil, "", err
}
```
(same with the later `else if`) because right now, if there _is_ an error, it gets swallowed and the program hits the `return nil, "", errors.New("unk")` statement.

Reply:
Not really, if there is an error parsing the first type of key, it will attempt parsing the second type of key. _Then_ if both failed to parse then it returns that "unk" error.

Reply:
That's a good point, you probably wanna just log the first error and try the second key type before returning an error. But ignoring errors makes me uneasy.

Reply:
FWIW I observed an `unk` message bar in Traffic Portal, testing bad certs.